### PR TITLE
Proper caching

### LINKED ISSUE #282: Implement cache: with proper push/pull to fix bug with caching in --shell-isolation
Implement proper cache syncing, instead of utilizing docker volume mounts.

Basicly copy files out if `policy:` is push-pull and copy files in if `policy:` is pull or push-pull.

Also implement globstar support while at it.

```yml
---
produce-cache:
  stage: build
  cache:
    key: global
    paths: [cache/**.txt]
  script: mkdir -p cache && touch cache/file1.txt

consume-cache:
  stage: test
  needs: [produce-cache]
  dependencies: [produce-cache]
  cache:
    key: global
    paths: [cache]
    policy: pull
  script: test -f cache/file1.txt
```

Remember to test for `Not uploading cache global due to policy` line...

### PR DESCRIPTION
**Pull request by firecow**
Implemented proper export/import functionality for cache:, added globstar support, and cache.policy support

CLOSES #282

### CODE REVIEW

**Comment by bcouetil** [comment]
I have some `exported cache` lines on some early npm jobs, but nothing of this sort on any maven job...

**Comment by bcouetil** [comment]
Ok

**Comment by bcouetil** [comment]
I'll try to find sometime tomorrow for this !

**Comment by deviantintegral** [comment]
I gave this a try, but rsync looks to be failing for me:

```
build_development > still running...
build_development exported cache vendor/ 'md-7830fd02c8e48425b6eb151dbe7e01d207e30f0a' in 9.72 s
build_development > still running...
build_development exported cache web/core 'md-7830fd02c8e48425b6eb151dbe7e01d207e30f0a' in 14 s
build_development exported cache web/modules/contrib 'md-7830fd02c8e48425b6eb151dbe7e01d207e30f0a' in 2.01 s
rsync: [sender] link_stat "/builds/build_development/web/themes/contrib" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1330) [sender=3.2.3]
```

That directory doesn't exist on disk, but gitlab CI itself happily skips it.

**Comment by firecow** [comment]
@deviantintegral @bcouetil I've added mutex functionality in https://github.com/firecow/gitlab-ci-local/pull/283/commits/a8732f91a3570aec384969c43612592c615801e9

Can you let me know, if this takes care of your issues?

I'm think parallel copy in/out could cause some race conditions.

**Comment by deviantintegral** [comment]
I'm still getting the above error due to specifying a cache key for a directory that doesn't exist.

**Comment by firecow** [comment]
> I'm still getting the above error due to specifying a cache key for a directory that doesn't exist.

Do you have a `.gitlab-ci.yml` I can use to reproduce @deviantintegral ?

**Comment by deviantintegral** [comment]
```yaml
image: webdevops/php-nginx:8.0-alpine

include:
  - project: 'dependabot-gitlab/dependabot-standalone'
    file: '.gitlab-ci.yml'

stages:
  - build
  - test
  - build_production

build_development:
  rules:
    - if: '$CI_PIPELINE_SOURCE == "merge_request_event"'
    - if: '$CI_COMMIT_BRANCH == "master"'
  stage: build
  cache:
    key:
      files:
        - composer.json
        - composer.lock
      prefix: $CI_JOB_NAME
    paths:
      - vendor/
      - web/core
      - web/modules/contrib
      - web/themes/contrib
  script:
    - composer install
  artifacts:
    paths:
      - vendor/
      - web/core/
      - web/modules/contrib
      - web/themes/contrib

install_drupal:
  rules:
    - if: '$CI_PIPELINE_SOURCE == "merge_request_event"'
    - if: '$CI_COMMIT_BRANCH == "master"'
  needs:
    - job: build_development
      artifacts: true
  stage: build
  services:
    - name: mariadb:10.4
      alias: db
  variables:
    MARIADB_USER: db
    MARIADB_PASSWORD: db
    MARIADB_DATABASE: db
    MARIADB_ROOT_PASSWORD: root
  script:
    - apk add --update mariadb-client
    - ./vendor/bin/drush site:install --existing-config --debug --verbose -y
    - ./vendor/bin/drush sql:dump > installed-drupal.sql
  artifacts:
    paths:
      - vendor/
      - web/core/
      - web/modules/contrib
      - web/themes/contrib
      - installed-drupal.sql
```

I'm running ```../../gitlab-ci-local/bin/macos/gitlab-ci-local --needs install_drupal```, and note that the `web/themes/contrib` directory doesn't exist in this repo yet.

**Comment by firecow** [comment]
@bcouetil Did you manage to verify that the Mutex commit fixed the bug? Performance issues aside that is.

**Comment by bcouetil** [comment]
I forgot sorry,

I'll have access to a decent network on tuesday.

**Comment by firecow** [comment]
@bcouetil I found a horrible mistake in the code you tested a while ago, that can explain the performance issue. There was a double copyCache out.
Also added `--mount-cache` option

**Comment by bcouetil** [comment]
OK thanks I'll run some test on this version tomorrow

**Comment by bcouetil** [comment]
Tests are running now, I'll have results tomorrow...

**Comment by bcouetil** [comment]
I conducted the same tests : 10 successives run of the same pipeline with the feature, then 10 successives with `--mount-cache`.

There still are same racing errors with the feature :cry: but less : 7 successful run out of 10.

Duration is still doubled, event a bit more, due to the single threaded synchronizations (I guess). It's a bit frustrating, because there is no need for any maven download on run 2 to 10 (the cache does not change a bit), and yet after the parallel jobs, we see long successive cache sync :

```
🔍✨-files-check-for-MR                         > still running...
🔍✨-files-check-for-MR                         exported cache .m2 'maven' in 16 s
🔍✨-files-check-for-MR                         finished in 34 s
🔍-Spring-Boot-version-check                   > still running...
🔍✨-gr-libs-check-for-MR                       > still running...
🏗-maven-package                               > still running...
🔍-Spring-Boot-version-check                   exported cache .m2 'maven' in 20 s
🔍-Spring-Boot-version-check                   finished in 44 s  WARN 1 
🔍✨-gr-libs-check-for-MR                       exported cache .m2 'maven' in 19 s
🔍✨-gr-libs-check-for-MR                       finished in 44 s
🏗-maven-package                               exported cache .m2 'maven' in 18 s
🏗-maven-package                               exported artifacts in 1.5 s
🏗-maven-package                               > still running...
🏗-maven-package                               copied artifacts to cwd in 330 ms
```

**Comment by firecow** [comment]
> I gave this a try, but rsync looks to be failing for me:
> 
> ```
> build_development > still running...
> build_development exported cache vendor/ 'md-7830fd02c8e48425b6eb151dbe7e01d207e30f0a' in 9.72 s
> build_development > still running...
> build_development exported cache web/core 'md-7830fd02c8e48425b6eb151dbe7e01d207e30f0a' in 14 s
> build_development exported cache web/modules/contrib 'md-7830fd02c8e48425b6eb151dbe7e01d207e30f0a' in 2.01 s
> rsync: [sender] link_stat "/builds/build_development/web/themes/contrib" failed: No such file or directory (2)
> rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1330) [sender=3.2.3]
> ```
> 
> That directory doesn't exist on disk, but gitlab CI itself happily skips it.

This should be fixed, btw :smile:

**Comment by firecow** [comment]
@bcouetil With `--mount-cache` all runs were successful and with no performance penalty, right?

**Comment by bcouetil** [comment]
Yes it works well, I compared quickly with the 4.24 version I had before, only a few seconds more, not sure why, I did not compare extensively.

Just an output problem, we see mounting logs both at start and at end of job, but I guess that there is no mounting at the end :

```
🔍🔢-version-check-and-calculate               mounting cache for path lib/node_modules
🔍🔢-version-check-and-calculate               mounting cache for path bin
[...]
🔍🔢-version-check-and-calculate               mounting cache for path lib/node_modules
🔍🔢-version-check-and-calculate               mounting cache for path bin
```

**Comment by firecow** [comment]
@bcouetil Well `/tmp` and `/var/lib/docker/volumes` are equal, its will only be beneficial to MacOS users...
Is there any chance I can get a hold of the ci/maven packages you are running, so I can do some proper debugging?

**Comment by firecow** [comment]
There is mounting both at the start, and at the end :smile: Btw

**Comment by bcouetil** [comment]
I think I can do better : building a minimal example failing !

**Comment by firecow** [comment]
@bcouetil Even better 😁

**Comment by bcouetil** [comment]
Here it is, with 0% success on my side :grin: 

Clone this repository : https://github.com/jeremylong/DependencyCheck

And use this :

```
.package:
  stage: build
  image: maven:3.6.3-jdk-11
  cache:
    key: maven
    paths:
      - .m2
    policy: pull-push
  script:
    - mvn package -DskipTests
  variables:
    MAVEN_OPTS: "-Dmaven.repo.local=.m2/repository"

package-1:
  extends: .package

package-2:
  extends: .package

package-3:
  extends: .package

package-4:
  extends: .package

package-5:
  extends: .package

package-6:
  extends: .package

package-7:
  extends: .package
```

**Comment by firecow** [comment]
@bcouetil Found the bug... async-mutex apparently doesn't work, but async-lock does...
`--mount-cache` also causes file conflicts once in a while, btw. Just not that often.

File: src/job.ts [hunk 1]
```diff
@@ -640,87 +669,118 @@ export class Job {
...
+            const time = process.hrtime();
+            const cacheName = await c.getUniqueCacheName(this.cwd, this.expandedVariables);
+            const cacheFolder = `${this.cwd}/.gitlab-ci-local/cache/${cacheName}`;
+            if (!await fs.pathExists(cacheFolder)) {
+                continue;
             }
-            return Utils.spawn(`docker cp ${folder}/. ${this._containerId}:/builds/${safeJobName}`);
-        };
+
+            await Mutex.waitForLock(cacheName);
```
Comment (FnuGk) [inline src/job.ts:686]: Should this not be a call to `await Mutex.exclusive(cacheName, async () => await this.copyIn(cacheFolder));` instead?
This just waits until no one has the mutex anymore and then carries on. What happens if someone grabs it right in the middle of the call to `copyIn` ?

Comment (firecow) [reply src/job.ts:686]: @FnuGk Hmmmm.... You are right...

Comment (firecow) [reply src/job.ts]: @FnuGk There you go

**Comment by firecow** [comment]
@FnuGk Can you try out this branch tomorrow?

**Review by FnuGk — APPROVED** [review]
If no one else has anything, then LGTM !

**Pull request merged**